sdk/dynamicconfig/metric/controller/push/period_matcher.go

```diff
@@ -51,26 +52,46 @@ func (matcher *PeriodMatcher) MarkStart(startTime time.Time) {
 // ApplySchedules sets the schedules that a PeriodMatcher consults when
 // constructing a Rule. After processing the schedules, ApplySchedules returns
 // the optimal period with which a controller should run a collection sweep.
+// If a period of 0 is returned, then metric collection should be halted.
 //
 // This function may be called concurrently.
-func (matcher *PeriodMatcher) ApplySchedules(sched []*pb.MetricConfigResponse_Schedule) time.Duration {
+func (matcher *PeriodMatcher) ApplySchedules(sched []*pb.MetricConfigResponse_Schedule) (time.Duration, error) {
+	if err := validate(sched); err != nil {
+		return 0, err
+	}
+
 	matcher.m.Lock()
 	matcher.sched = sched
 	matcher.metrics = make(map[string]*collectData)
 	matcher.m.Unlock()
 
-	return getExportPeriod(matcher.sched)
+	return getExportPeriod(matcher.sched), nil
```

Comment:
getExportPeriods(sched) can happen before grabbing the lock.

exportPeriod, err := getExportPeriod(sche)
if err != nil:
  return 0, err

matcher.m.Lock()
....
return exportPeriod, nil

Reply:
Ahhh I see, fixing